File: plugin-core/src/main/java/com/github/catatafishen/agentbridge/settings/ToolsConfigurable.java
package com.github.catatafishen.agentbridge.settings;

import com.github.catatafishen.agentbridge.services.ToolDefinition;
import com.github.catatafishen.agentbridge.services.ToolRegistry;
import com.github.catatafishen.agentbridge.ui.ThemeColor;
import com.github.catatafishen.agentbridge.ui.ToolKindColors;
import com.intellij.openapi.options.Configurable;
import com.intellij.openapi.project.Project;
import com.intellij.ui.TitledSeparator;
import com.intellij.ui.components.JBCheckBox;
import com.intellij.ui.components.JBLabel;
import com.intellij.ui.components.JBPanel;
import com.intellij.ui.components.JBScrollPane;
import com.intellij.util.ui.JBUI;
import com.intellij.util.ui.UIUtil;
import org.jetbrains.annotations.Nls;
import org.jetbrains.annotations.NotNull;
import org.jetbrains.annotations.Nullable;

import javax.swing.*;
import java.awt.*;
import java.util.ArrayList;
import java.util.LinkedHashMap;
import java.util.List;
import java.util.Map;
import java.util.Objects;

/**
 * Settings page: Settings → Tools → AgentBridge → MCP → Tools.
 * Enable or disable individual MCP tools exposed to agents, and customize per-kind colors.
 */
public final class ToolsConfigurable implements Configurable {

    private final Project project;
    private final Map<String, JBCheckBox> toolCheckboxes = new LinkedHashMap<>();
    /**
     * Per-category lists of checkboxes, used for section-level enable/disable.
     */
    private final Map<ToolRegistry.Category, List<JBCheckBox>> categoryCheckboxes = new LinkedHashMap<>();

    private @Nullable JBLabel counterLabel;
    private @Nullable ThemeColorComboBox readColorCombo;
    private @Nullable ThemeColorComboBox searchColorCombo;
    private @Nullable ThemeColorComboBox editColorCombo;
    private @Nullable ThemeColorComboBox executeColorCombo;

    public ToolsConfigurable(@NotNull Project project) {
        this.project = project;
    }

    @Nls
    @Override
    public String getDisplayName() {
        return "Tools";
    }

    @NotNull
    @Override
    public JComponent createComponent() {
        McpServerSettings settings = McpServerSettings.getInstance(project);

        JBPanel<?> toolsPanel = new JBPanel<>();
        toolsPanel.setLayout(new BoxLayout(toolsPanel, BoxLayout.Y_AXIS));
        toolsPanel.setBorder(JBUI.Borders.empty(8));

        // Tool counter and limit hint
        counterLabel = new JBLabel();
        counterLabel.setFont(JBUI.Fonts.label());
        counterLabel.setAlignmentX(Component.LEFT_ALIGNMENT);
        counterLabel.setBorder(JBUI.Borders.empty(0, 0, 6, 0));
        toolsPanel.add(counterLabel);

        JBLabel limitHint = new JBLabel(
            "<html>MCP clients enforce a <b>" + McpToolFilter.MAX_TOOLS
                + "-tool limit</b>. Only enable tools you intend to use.</html>");
        limitHint.setFont(JBUI.Fonts.smallFont());
        limitHint.setForeground(UIUtil.getContextHelpForeground());
        limitHint.setBorder(JBUI.Borders.empty(0, 0, 6, 0));
        limitHint.setAlignmentX(Component.LEFT_ALIGNMENT);
        toolsPanel.add(limitHint);

        // Global enable/disable row at top
        JButton enableAllBtn = new JButton("Enable All");
        JButton disableAllBtn = new JButton("Disable All");
        enableAllBtn.addActionListener(e -> enableUpToMax(toolCheckboxes.values()));
        disableAllBtn.addActionListener(e -> {
            toolCheckboxes.values().forEach(cb -> cb.setSelected(false));
            updateCounter();
        });

        JBPanel<?> topRow = new JBPanel<>();
        topRow.setLayout(new BoxLayout(topRow, BoxLayout.X_AXIS));
        topRow.setBorder(JBUI.Borders.empty(0, 0, 4, 0));
        topRow.add(enableAllBtn);
        topRow.add(Box.createHorizontalStrut(JBUI.scale(8)));
        topRow.add(disableAllBtn);
        topRow.add(Box.createHorizontalGlue());
        topRow.setAlignmentX(Component.LEFT_ALIGNMENT);
        toolsPanel.add(topRow);

        // Tool list grouped by category
        List<ToolDefinition> tools = McpToolFilter.getConfigurableTools(project);
        ToolRegistry.Category currentCategory = null;

        for (ToolDefinition tool : tools) {
            if (tool.category() != currentCategory) {
                currentCategory = tool.category();
                final ToolRegistry.Category cat = currentCategory;
                toolsPanel.add(buildCategoryHeader(cat));
            }

            Color kindColor = kindColorFor(tool, settings);
            JBPanel<?> toolRow = new JBPanel<>();
            toolRow.setLayout(new BoxLayout(toolRow, BoxLayout.X_AXIS));
            toolRow.setAlignmentX(Component.LEFT_ALIGNMENT);
            toolRow.setMaximumSize(new Dimension(Integer.MAX_VALUE, toolRow.getPreferredSize().height));

            // Colored kind dot
            JBLabel dot = new JBLabel("● ");
            dot.setForeground(kindColor);
            dot.setBorder(JBUI.Borders.empty(1, 16, 0, 0));
            toolRow.add(dot);

            JBCheckBox cb = new JBCheckBox(tool.displayName(), settings.isToolEnabled(tool.id()));
            cb.setBorder(JBUI.Borders.empty(1, 0, 0, 0));
            cb.addItemListener(e -> onCheckboxToggled(cb));
            toolCheckboxes.put(tool.id(), cb);
            categoryCheckboxes.computeIfAbsent(tool.category(), k -> new ArrayList<>()).add(cb);
            toolRow.add(cb);
            toolRow.add(Box.createHorizontalGlue());

            toolsPanel.add(toolRow);

            String desc = tool.description();
            if (!desc.isBlank()) {
                // Use HTML so the label wraps when the panel is width-constrained.
                // Preferred width is set to 1 so this label does not drive the panel's
                // preferred width — the wrapping container determines the actual width.
                JBLabel descLabel = new JBLabel("<html>" + desc + "</html>");
                descLabel.setFont(descLabel.getFont().deriveFont((float) (JBUI.Fonts.label().getSize() - 1)));
                descLabel.setForeground(UIUtil.getContextHelpForeground());
                descLabel.setBorder(JBUI.Borders.empty(0, 36, 3, 0));
                descLabel.setAlignmentX(Component.LEFT_ALIGNMENT);
                descLabel.setPreferredSize(new Dimension(1, descLabel.getPreferredSize().height));
                descLabel.setMaximumSize(new Dimension(Integer.MAX_VALUE, descLabel.getPreferredSize().height));
                toolsPanel.add(descLabel);
            }
        }

        // Color picker section
        toolsPanel.add(buildColorPickerSection(settings));
        toolsPanel.add(Box.createVerticalGlue());

        // Wrap in a BorderLayout panel so the scroll pane forces the content width.
        // toolsPanel in NORTH gets the full viewport width, preventing long labels from
        // widening the panel and pushing the Enable/Disable buttons off-screen.
        JBPanel<?> scrollContent = new JBPanel<>(new BorderLayout());
        scrollContent.add(toolsPanel, BorderLayout.NORTH);

        JBScrollPane scrollPane = new JBScrollPane(scrollContent);
        scrollPane.setBorder(JBUI.Borders.empty());
        scrollPane.setHorizontalScrollBarPolicy(ScrollPaneConstants.HORIZONTAL_SCROLLBAR_NEVER);

        JBPanel<?> wrapper = new JBPanel<>(new BorderLayout());
        wrapper.add(scrollPane, BorderLayout.CENTER);

        updateCounter();
        return wrapper;
    }

    private int countEnabled() {
        return (int) toolCheckboxes.values().stream().filter(JBCheckBox::isSelected).count();
    }

    private void updateCounter() {
        if (counterLabel == null) return;
        int enabled = countEnabled();
        int max = McpToolFilter.MAX_TOOLS;
        boolean overLimit = enabled > max;
        counterLabel.setText(enabled + " / " + max + " tools enabled"
            + (overLimit ? "  ⚠ over limit!" : ""));
        counterLabel.setForeground(overLimit
            ? UIUtil.getErrorForeground()
            : UIUtil.getLabelForeground());
    }

    /**
     * Called when a tool checkbox is toggled. If enabling would exceed the limit,
     * revert the selection.
     */
    private void onCheckboxToggled(JBCheckBox cb) {
        if (cb.isSelected() && countEnabled() > McpToolFilter.MAX_TOOLS) {
            cb.setSelected(false);
            return;
        }
        updateCounter();
    }

    /**
     * Enables checkboxes in iteration order until the global limit is reached.
     * Already-enabled checkboxes count towards the limit but are not toggled off.
     */
    private void enableUpToMax(java.util.Collection<JBCheckBox> checkboxes) {
        int enabled = countEnabled();
        for (JBCheckBox cb : checkboxes) {
            if (cb.isSelected()) continue;
            if (enabled >= McpToolFilter.MAX_TOOLS) break;
            cb.setSelected(true);
            enabled++;
        }
        updateCounter();
    }

    /**
     * Builds a category header row: separator on the left, Enable/Disable buttons on the right.
     */
    private JComponent buildCategoryHeader(ToolRegistry.Category category) {
        JBPanel<?> row = new JBPanel<>(new BorderLayout());
        row.setAlignmentX(Component.LEFT_ALIGNMENT);
        row.setBorder(JBUI.Borders.empty(8, 0, 2, 0));

        TitledSeparator sep = new TitledSeparator(category.displayName);
        row.add(sep, BorderLayout.CENTER);

        JButton sectionEnableBtn = new JButton("Enable");
        JButton sectionDisableBtn = new JButton("Disable");
        sectionEnableBtn.setFont(JBUI.Fonts.smallFont());
        sectionDisableBtn.setFont(JBUI.Fonts.smallFont());

        sectionEnableBtn.addActionListener(e -> {
            List<JBCheckBox> cbs = categoryCheckboxes.get(category);
            if (cbs != null) enableUpToMax(cbs);
        });
        sectionDisableBtn.addActionListener(e -> {
            List<JBCheckBox> cbs = categoryCheckboxes.get(category);
            if (cbs != null) {
                cbs.forEach(cb -> cb.setSelected(false));
                updateCounter();
            }
        });

        JBPanel<?> btnRow = new JBPanel<>();
        btnRow.setLayout(new BoxLayout(btnRow, BoxLayout.X_AXIS));
        btnRow.add(sectionEnableBtn);
        btnRow.add(Box.createHorizontalStrut(JBUI.scale(4)));
        btnRow.add(sectionDisableBtn);
        row.add(btnRow, BorderLayout.EAST);

        return row;
    }

    /**
     * Builds the "Tool Kind Colors" section with a {@link ThemeColorComboBox} for each kind.
     */
    private JComponent buildColorPickerSection(McpServerSettings settings) {
        JBPanel<?> section = new JBPanel<>();
        section.setLayout(new BoxLayout(section, BoxLayout.Y_AXIS));
        section.setAlignmentX(Component.LEFT_ALIGNMENT);

        TitledSeparator colorSep = new TitledSeparator("Tool Kind Colors");
        colorSep.setBorder(JBUI.Borders.empty(16, 0, 4, 0));
        colorSep.setAlignmentX(Component.LEFT_ALIGNMENT);
        section.add(colorSep);

        JBLabel hint = new JBLabel(
            "<html>Customize the accent color used for each tool kind in this settings panel, "
                + "tool-chip labels in the chat view, and permission dropdowns. "
                + "Colors adapt automatically when the IDE theme changes.</html>");
        hint.setFont(JBUI.Fonts.smallFont());
        hint.setForeground(UIUtil.getContextHelpForeground());
        hint.setBorder(JBUI.Borders.empty(0, 0, 8, 0));
        hint.setAlignmentX(Component.LEFT_ALIGNMENT);
        hint.setPreferredSize(new Dimension(1, hint.getPreferredSize().height));
        hint.setMaximumSize(new Dimension(Integer.MAX_VALUE, hint.getPreferredSize().height));
        section.add(hint);

        readColorCombo = new ThemeColorComboBox();
        searchColorCombo = new ThemeColorComboBox();
        editColorCombo = new ThemeColorComboBox();
        executeColorCombo = new ThemeColorComboBox();

        readColorCombo.setSelectedThemeColor(ThemeColor.fromKey(settings.getKindReadColorKey()));
        searchColorCombo.setSelectedThemeColor(ThemeColor.fromKey(settings.getKindSearchColorKey()));
        editColorCombo.setSelectedThemeColor(ThemeColor.fromKey(settings.getKindEditColorKey()));
        executeColorCombo.setSelectedThemeColor(ThemeColor.fromKey(settings.getKindExecuteColorKey()));

        section.add(colorRow("Read & Navigate", readColorCombo));
        section.add(colorRow("Search & Query", searchColorCombo));
        section.add(colorRow("Edit & Refactor", editColorCombo));
        section.add(colorRow("Run & Execute", executeColorCombo));

        return section;
    }

    private static JComponent colorRow(String label, ThemeColorComboBox combo) {
        JBPanel<?> row = new JBPanel<>();
        row.setLayout(new BoxLayout(row, BoxLayout.X_AXIS));
        row.setAlignmentX(Component.LEFT_ALIGNMENT);
        row.setBorder(JBUI.Borders.empty(2, 0));
        row.setMaximumSize(new Dimension(Integer.MAX_VALUE, row.getPreferredSize().height));

        JBLabel lbl = new JBLabel(label);
        lbl.setPreferredSize(new Dimension(JBUI.scale(140), lbl.getPreferredSize().height));
        lbl.setMaximumSize(new Dimension(JBUI.scale(140), lbl.getPreferredSize().height));
        row.add(lbl);
        combo.setMaximumSize(new Dimension(JBUI.scale(180), combo.getPreferredSize().height));
        row.add(combo);
        row.add(Box.createHorizontalGlue());
        return row;
    }

    /**
     * Returns the kind accent color for [tool], honoring settings overrides.
     */
    private static Color kindColorFor(ToolDefinition tool, McpServerSettings settings) {
        return switch (tool.kind()) {
            case SEARCH -> ToolKindColors.searchColor(settings);
            case EDIT, WRITE, DELETE, MOVE -> ToolKindColors.editColor(settings);
            case EXECUTE -> ToolKindColors.executeColor(settings);
            default -> ToolKindColors.readColor(settings);
        };
    }

    @Override
    public boolean isModified() {
        McpServerSettings settings = McpServerSettings.getInstance(project);
        for (Map.Entry<String, JBCheckBox> entry : toolCheckboxes.entrySet()) {
            if (entry.getValue().isSelected() != settings.isToolEnabled(entry.getKey())) return true;
        }
        if (readColorCombo != null
            && !Objects.equals(keyOf(readColorCombo), settings.getKindReadColorKey())) return true;
        if (searchColorCombo != null
            && !Objects.equals(keyOf(searchColorCombo), settings.getKindSearchColorKey())) return true;
        if (editColorCombo != null
            && !Objects.equals(keyOf(editColorCombo), settings.getKindEditColorKey())) return true;
        return executeColorCombo != null
            && !Objects.equals(keyOf(executeColorCombo), settings.getKindExecuteColorKey());
    }

    @Override
    public void apply() {
        McpServerSettings settings = McpServerSettings.getInstance(project);
        for (Map.Entry<String, JBCheckBox> entry : toolCheckboxes.entrySet()) {
            settings.setToolEnabled(entry.getKey(), entry.getValue().isSelected());
        }
        if (readColorCombo != null) settings.setKindReadColorKey(keyOf(readColorCombo));
        if (searchColorCombo != null) settings.setKindSearchColorKey(keyOf(searchColorCombo));
        if (editColorCombo != null) settings.setKindEditColorKey(keyOf(editColorCombo));
        if (executeColorCombo != null) settings.setKindExecuteColorKey(keyOf(executeColorCombo));
    }

    @Override
    public void reset() {
        McpServerSettings settings = McpServerSettings.getInstance(project);
        for (Map.Entry<String, JBCheckBox> entry : toolCheckboxes.entrySet()) {
            entry.getValue().setSelected(settings.isToolEnabled(entry.getKey()));
        }
        if (readColorCombo != null)
            readColorCombo.setSelectedThemeColor(ThemeColor.fromKey(settings.getKindReadColorKey()));
        if (searchColorCombo != null)
            searchColorCombo.setSelectedThemeColor(ThemeColor.fromKey(settings.getKindSearchColorKey()));
        if (editColorCombo != null)
            editColorCombo.setSelectedThemeColor(ThemeColor.fromKey(settings.getKindEditColorKey()));
        if (executeColorCombo != null)
            executeColorCombo.setSelectedThemeColor(ThemeColor.fromKey(settings.getKindExecuteColorKey()));
        updateCounter();
    }

    @Override
    public void disposeUIResources() {
        counterLabel = null;
        readColorCombo = null;
        searchColorCombo = null;
        editColorCombo = null;
        executeColorCombo = null;
        toolCheckboxes.clear();
        categoryCheckboxes.clear();
    }

    /**
     * Returns the ThemeColor name for persistence, or null if "Default" is selected.
     */
    private static @Nullable String keyOf(ThemeColorComboBox combo) {
        ThemeColor tc = combo.getSelectedThemeColor();
        return tc != null ? tc.name() : null;
    }
}

File: plugin-core/src/main/java/com/github/catatafishen/agentbridge/settings/ToolsConfigurable.kt
package com.github.catatafishen.agentbridge.settings

import com.github.catatafishen.agentbridge.psi.PlatformApiCompat
import com.github.catatafishen.agentbridge.psi.PsiBridgeService
import com.github.catatafishen.agentbridge.psi.tools.rider.ReSharperMcpClient
import com.github.catatafishen.agentbridge.services.ToolDefinition
import com.github.catatafishen.agentbridge.services.ToolRegistry
import com.github.catatafishen.agentbridge.ui.ThemeColor
import com.github.catatafishen.agentbridge.ui.ToolKindColors
import com.intellij.ide.BrowserUtil
import com.intellij.openapi.options.BoundConfigurable
import com.intellij.openapi.options.SearchableConfigurable
import com.intellij.openapi.project.Project
import com.intellij.ui.TitledSeparator
import com.intellij.ui.components.JBCheckBox
import com.intellij.ui.components.JBLabel
import com.intellij.ui.components.JBPanel
import com.intellij.ui.components.JBScrollPane
import com.intellij.ui.dsl.builder.AlignX
import com.intellij.ui.dsl.builder.AlignY
import com.intellij.ui.dsl.builder.RowLayout
import com.intellij.ui.dsl.builder.panel
import com.intellij.util.ui.JBUI
import com.intellij.util.ui.UIUtil
import java.awt.BorderLayout
import java.awt.Color
import java.awt.Component
import java.awt.Dimension
import java.awt.event.ComponentAdapter
import java.awt.event.ComponentEvent
import java.util.*
import java.util.concurrent.ExecutionException
import javax.swing.*
import javax.swing.event.HyperlinkEvent

class ToolsConfigurable(private val project: Project) :
    BoundConfigurable("Tools"),
    SearchableConfigurable {

    override fun getId(): String = "com.github.catatafishen.agentbridge.tools"

    private val toolCheckboxes = LinkedHashMap<String, JBCheckBox>()
    private val categoryCheckboxes = LinkedHashMap<ToolRegistry.Category, MutableList<JBCheckBox>>()

    private var counterLabel: JBLabel? = null
    private var readColorCombo: ThemeColorComboBox? = null
    private var searchColorCombo: ThemeColorComboBox? = null
    private var editColorCombo: ThemeColorComboBox? = null
    private var executeColorCombo: ThemeColorComboBox? = null

    override fun createPanel() = panel {
        row {
            cell(buildToolsContent())
                .align(AlignX.FILL).align(AlignY.FILL).resizableColumn()
        }.resizableRow().layout(RowLayout.PARENT_GRID)
        onIsModified { computeIsModified() }
        onApply { applySettings() }
        onReset { resetFromSettings() }
    }

    private fun buildToolsContent(): JComponent {
        val settings = McpServerSettings.getInstance(project)

        val toolsPanel = JBPanel<JBPanel<*>>().apply {
            layout = BoxLayout(this, BoxLayout.Y_AXIS)
            border = JBUI.Borders.empty(8)
        }

        counterLabel = JBLabel().apply {
            alignmentX = Component.LEFT_ALIGNMENT
            border = JBUI.Borders.emptyBottom(4)
        }
        toolsPanel.add(counterLabel)

        val limitHint = JBLabel(
            "<html>MCP clients enforce a <b>${McpToolFilter.MAX_TOOLS}-tool limit</b>. " +
                "Only enable tools you intend to use.</html>"
        ).apply {
            font = JBUI.Fonts.smallFont()
            foreground = UIUtil.getContextHelpForeground()
            border = JBUI.Borders.emptyBottom(6)
            alignmentX = Component.LEFT_ALIGNMENT
            isAllowAutoWrapping = true
        }
        toolsPanel.add(limitHint)

        if (PlatformApiCompat.isPluginInstalled("com.intellij.modules.rider")) {
            val container = JBPanel<JBPanel<*>>().apply {
                layout = BoxLayout(this, BoxLayout.Y_AXIS)
                alignmentX = Component.LEFT_ALIGNMENT
            }
            toolsPanel.add(container)
            object : SwingWorker<Boolean, Unit>() {
                override fun doInBackground(): Boolean = ReSharperMcpClient.isAvailable()
                override fun done() {
                    var available = false
                    try {
                        available = get()
                    } catch (_: InterruptedException) {
                        Thread.currentThread().interrupt()
                    } catch (_: ExecutionException) {
                        // probe failed; treat as unavailable
                    }
                    if (!available) {
                        container.add(buildRiderInfoPanel())
                        container.revalidate(); container.repaint()
                    }
                }
            }.execute()
        }

        val enableAllBtn = JButton("Enable All").apply {
            addActionListener {
                toolCheckboxes.values.forEach { it.isSelected = true }
                updateCounter()
            }
        }
        val disableAllBtn = JButton("Disable All").apply {
            addActionListener {
                toolCheckboxes.values.forEach { it.isSelected = false }
                updateCounter()
            }
        }
        val topRow = JBPanel<JBPanel<*>>().apply {
            layout = BoxLayout(this, BoxLayout.X_AXIS)
            border = JBUI.Borders.emptyBottom(4)
            add(enableAllBtn)
            add(Box.createHorizontalStrut(JBUI.scale(8)))
            add(disableAllBtn)
            add(Box.createHorizontalGlue())
            alignmentX = Component.LEFT_ALIGNMENT
            maximumSize = Dimension(Int.MAX_VALUE, preferredSize.height)
        }
        toolsPanel.add(topRow)

        addToolRows(toolsPanel, settings)

        toolsPanel.add(buildColorPickerSection(settings))
        toolsPanel.add(Box.createVerticalGlue())

        // Revalidate when the panel's width changes. BoxLayout commits to each child's
        // preferredSize.height before layout assigns them a real width. JBLabel with
        // isAllowAutoWrapping=true can only recalculate its wrapped height once getWidth()>0,
        // so we need a second layout pass after the first width arrives.
        toolsPanel.addComponentListener(object : ComponentAdapter() {
            private var lastWidth = -1
            override fun componentResized(e: ComponentEvent) {
                val w = toolsPanel.width
                if (w > 0 && w != lastWidth) {
                    lastWidth = w
                    SwingUtilities.invokeLater {
                        toolsPanel.revalidate()
                        toolsPanel.repaint()
                    }
                }
            }
        })

        val scrollContent = JBPanel<JBPanel<*>>(BorderLayout()).apply {
            add(toolsPanel, BorderLayout.NORTH)
        }
        val scrollPane = JBScrollPane(scrollContent).apply {
            border = JBUI.Borders.empty()
            horizontalScrollBarPolicy = ScrollPaneConstants.HORIZONTAL_SCROLLBAR_NEVER
            // Cap preferred width so the Settings dialog does not grow to fit the
            // full single-line text width of tool descriptions. AlignX.FILL in
            // createPanel() will expand the pane to the actual dialog width anyway;
            // this only controls what the dialog uses when computing its own preferred size.
            preferredSize = Dimension(JBUI.scale(580), JBUI.scale(480))
        }
        updateCounter()
        return scrollPane
    }

    private fun addToolRows(toolsPanel: JPanel, settings: McpServerSettings) {
        var currentCategory: ToolRegistry.Category? = null
        for (tool in sortedConfigurableTools()) {
            val category = tool.category()
            if (category != currentCategory) {
                currentCategory = category
                toolsPanel.add(buildCategoryHeader(category))
            }
            addToolRow(toolsPanel, tool, category, settings)
        }
    }

    private fun sortedConfigurableTools(): List<ToolDefinition> =
        McpToolFilter.getConfigurableTools(project)
            .sortedWith(
                compareBy<ToolDefinition> { it.category().displayName.lowercase(Locale.ROOT) }
                    .thenBy { it.displayName().lowercase(Locale.ROOT) }
                    .thenBy { it.id() }
            )

    private fun addToolRow(
        toolsPanel: JPanel,
        tool: ToolDefinition,
        category: ToolRegistry.Category,
        settings: McpServerSettings
    ) {
        val kindColor = kindColorFor(tool, settings)
        val toolRow = JBPanel<JBPanel<*>>().apply {
            layout = BoxLayout(this, BoxLayout.X_AXIS)
            alignmentX = Component.LEFT_ALIGNMENT
            maximumSize = Dimension(Int.MAX_VALUE, preferredSize.height)
        }
        val dot = JBLabel("● ").apply {
            foreground = kindColor
            border = JBUI.Borders.empty(1, 16, 0, 0)
        }
        toolRow.add(dot)
        val cb = JBCheckBox(tool.displayName(), settings.isToolEnabled(tool.id())).apply {
            border = JBUI.Borders.emptyTop(1)
            addItemListener { updateCounter() }
        }
        toolCheckboxes[tool.id()] = cb
        categoryCheckboxes.getOrPut(category) { mutableListOf() }.add(cb)
        toolRow.add(cb)
        toolRow.add(Box.createHorizontalGlue())
        toolsPanel.add(toolRow)
        addToolDescription(toolsPanel, tool)
    }

    private fun addToolDescription(toolsPanel: JPanel, tool: ToolDefinition) {
        val desc = tool.description()
        if (desc.isBlank()) return
        val descLabel = JBLabel("<html>$desc</html>").apply {
            font = font.deriveFont((JBUI.Fonts.label().size - 1).toFloat())
            foreground = UIUtil.getContextHelpForeground()
            border = JBUI.Borders.empty(0, 36, 3, 0)
            alignmentX = Component.LEFT_ALIGNMENT
            isAllowAutoWrapping = true
        }
        toolsPanel.add(descLabel)
    }

    private fun countEnabled(): Int = toolCheckboxes.values.count { it.isSelected }

    private fun updateCounter() {
        val label = counterLabel ?: return
        val enabled = countEnabled()
        val max = McpToolFilter.MAX_TOOLS
        val overLimit = enabled > max
        label.text =
            "<html><b>$enabled / $max tools enabled</b>${if (overLimit) "  ⚠ over limit!" else ""}</html>"
        label.foreground = if (overLimit) UIUtil.getErrorForeground() else UIUtil.getLabelForeground()
    }

    private fun buildCategoryHeader(category: ToolRegistry.Category): JComponent {
        val row = JBPanel<JBPanel<*>>(BorderLayout()).apply {
            alignmentX = Component.LEFT_ALIGNMENT
            border = JBUI.Borders.empty(8, 0, 2, 0)
        }
        row.add(TitledSeparator(category.displayName), BorderLayout.CENTER)

        val sectionEnableBtn = JButton("Enable").apply {
            font = JBUI.Fonts.smallFont()
            addActionListener {
                categoryCheckboxes[category]?.forEach { it.isSelected = true }
                updateCounter()
            }
        }
        val sectionDisableBtn = JButton("Disable").apply {
            font = JBUI.Fonts.smallFont()
            addActionListener {
                categoryCheckboxes[category]?.forEach { it.isSelected = false }
                updateCounter()
            }
        }
        val btnRow = JBPanel<JBPanel<*>>().apply {
            layout = BoxLayout(this, BoxLayout.X_AXIS)
            add(sectionEnableBtn)
            add(Box.createHorizontalStrut(JBUI.scale(4)))
            add(sectionDisableBtn)
        }
        row.add(btnRow, BorderLayout.EAST)
        return row
    }

    private fun buildColorPickerSection(settings: McpServerSettings): JComponent {
        val section = JBPanel<JBPanel<*>>().apply {
            layout = BoxLayout(this, BoxLayout.Y_AXIS)
            alignmentX = Component.LEFT_ALIGNMENT
        }
        val sep = TitledSeparator("Tool Kind Colors").apply {
            border = JBUI.Borders.empty(16, 0, 4, 0)
            alignmentX = Component.LEFT_ALIGNMENT
        }
        section.add(sep)
        val hint = JBLabel(
            "<html>Customize the accent color used for each tool kind in this settings panel, " +
                "tool-chip labels in the chat view, and permission dropdowns. " +
                "Colors adapt automatically when the IDE theme changes.</html>"
        ).apply {
            font = JBUI.Fonts.smallFont()
            foreground = UIUtil.getContextHelpForeground()
            border = JBUI.Borders.emptyBottom(8)
            alignmentX = Component.LEFT_ALIGNMENT
            isAllowAutoWrapping = true
        }
        section.add(hint)

        readColorCombo = ThemeColorComboBox().also {
            it.selectedThemeColor = ThemeColor.fromKey(settings.kindReadColorKey)
        }
        searchColorCombo = ThemeColorComboBox().also {
            it.selectedThemeColor = ThemeColor.fromKey(settings.kindSearchColorKey)
        }
        editColorCombo = ThemeColorComboBox().also {
            it.selectedThemeColor = ThemeColor.fromKey(settings.kindEditColorKey)
        }
        executeColorCombo = ThemeColorComboBox().also {
            it.selectedThemeColor = ThemeColor.fromKey(settings.kindExecuteColorKey)
        }
        section.add(colorRow("Read & Navigate", readColorCombo!!))
        section.add(colorRow("Search & Query", searchColorCombo!!))
        section.add(colorRow("Edit & Refactor", editColorCombo!!))
        section.add(colorRow("Run & Execute", executeColorCombo!!))
        return section
    }

    private fun colorRow(label: String, combo: ThemeColorComboBox): JComponent {
        val row = JBPanel<JBPanel<*>>().apply {
            layout = BoxLayout(this, BoxLayout.X_AXIS)
            alignmentX = Component.LEFT_ALIGNMENT
            border = JBUI.Borders.empty(2, 0)
            maximumSize = Dimension(Int.MAX_VALUE, preferredSize.height)
        }
        val lbl = JBLabel(label).apply {
            preferredSize = Dimension(JBUI.scale(140), preferredSize.height)
            maximumSize = Dimension(JBUI.scale(140), preferredSize.height)
        }
        row.add(lbl)
        combo.maximumSize = Dimension(JBUI.scale(180), combo.preferredSize.height)
        row.add(combo)
        row.add(Box.createHorizontalGlue())
        return row
    }

    private fun kindColorFor(tool: ToolDefinition, settings: McpServerSettings): Color =
        when (tool.kind()) {
            ToolDefinition.Kind.SEARCH -> ToolKindColors.searchColor(settings)
            ToolDefinition.Kind.EDIT, ToolDefinition.Kind.WRITE,
            ToolDefinition.Kind.DELETE, ToolDefinition.Kind.MOVE -> ToolKindColors.editColor(settings)

            ToolDefinition.Kind.EXECUTE -> ToolKindColors.executeColor(settings)
            else -> ToolKindColors.readColor(settings)
        }

    private fun computeIsModified(): Boolean {
        val settings = McpServerSettings.getInstance(project)
        for ((id, cb) in toolCheckboxes) {
            if (cb.isSelected != settings.isToolEnabled(id)) return true
        }
        if (readColorCombo != null && keyOf(readColorCombo!!) != settings.kindReadColorKey) return true
        if (searchColorCombo != null && keyOf(searchColorCombo!!) != settings.kindSearchColorKey) return true
        if (editColorCombo != null && keyOf(editColorCombo!!) != settings.kindEditColorKey) return true
        return executeColorCombo != null && keyOf(executeColorCombo!!) != settings.kindExecuteColorKey
    }

    private fun applySettings() {
        val settings = McpServerSettings.getInstance(project)
        for ((id, cb) in toolCheckboxes) settings.setToolEnabled(id, cb.isSelected)
        readColorCombo?.let { settings.kindReadColorKey = keyOf(it) }
        searchColorCombo?.let { settings.kindSearchColorKey = keyOf(it) }
        editColorCombo?.let { settings.kindEditColorKey = keyOf(it) }
        executeColorCombo?.let { settings.kindExecuteColorKey = keyOf(it) }
    }

    private fun resetFromSettings() {
        val settings = McpServerSettings.getInstance(project)
        for ((id, cb) in toolCheckboxes) cb.isSelected = settings.isToolEnabled(id)
        readColorCombo?.selectedThemeColor = ThemeColor.fromKey(settings.kindReadColorKey)
        searchColorCombo?.selectedThemeColor = ThemeColor.fromKey(settings.kindSearchColorKey)
        editColorCombo?.selectedThemeColor = ThemeColor.fromKey(settings.kindEditColorKey)
        executeColorCombo?.selectedThemeColor = ThemeColor.fromKey(settings.kindExecuteColorKey)
        updateCounter()
    }

    override fun disposeUIResources() {
        super<BoundConfigurable>.disposeUIResources()
        counterLabel = null
        readColorCombo = null
        searchColorCombo = null
        editColorCombo = null
        executeColorCombo = null
        toolCheckboxes.clear()
        categoryCheckboxes.clear()
    }

    private fun keyOf(combo: ThemeColorComboBox): String? = combo.selectedThemeColor?.name

    private fun buildRiderInfoPanel(): JComponent {
        val disabledList = PsiBridgeService.getRiderDisabledToolIds().joinToString(", ")
        val pane = JEditorPane(
            "text/html",
            "<html><body><b>⚠ Rider:</b> The following tools are unavailable in Rider without the " +
                "<a href='https://github.com/joshua-light/resharper-mcp'>resharper-mcp</a> " +
                "companion plugin: <br><i>$disabledList</i></body></html>"
        ).apply {
            isEditable = false
            isOpaque = false
            foreground = UIUtil.getContextHelpForeground()
            alignmentX = Component.LEFT_ALIGNMENT
            border = JBUI.Borders.empty(4, 0, 8, 0)
            maximumSize = Dimension(Int.MAX_VALUE, Int.MAX_VALUE)
            addHyperlinkListener { e ->
                if (HyperlinkEvent.EventType.ACTIVATED == e.eventType && e.url != null) {
                    BrowserUtil.browse(e.url.toExternalForm())
                }
            }
        }
        return pane
    }
}
